File: cpp/img_to_cam.cpp
#include <iostream>
#include <opencv2/opencv.hpp>

cv::Mat pointImgToCam(const cv::Mat &K, const cv::Mat &R, const cv::Mat &t, const cv::Point2f &P_img, double Z) {
    // 构建[R|t]矩阵
    cv::Mat Rt(3, 4, CV_64F);
    R.copyTo(Rt(cv::Rect(0, 0, 3, 3)));
    t.copyTo(Rt(cv::Rect(3, 0, 1, 3)));

    // 计算[R|t]和K的逆矩阵
    cv::Mat K_inv = K.inv();
    cv::Mat Rt_inv;
    cv::invert(Rt(cv::Rect(0, 0, 3, 3)), Rt_inv);  // 仅取旋转矩阵部分求逆

    // 构建图像坐标的齐次坐标
    cv::Mat P_img_homo = (cv::Mat_<double>(3, 1) << P_img.x * Z, P_img.y * Z, Z);

    // 计算相机坐标系下的坐标
    cv::Mat P_cam = Rt_inv * (K_inv * P_img_homo - Rt(cv::Rect(3, 0, 1, 3)));
    
    return P_cam;
}

int main() {
    cv::Mat K = (cv::Mat_<double>(3, 3) << 700, 0, 300, 0, 700, 200, 0, 0, 1);
    cv::Mat R = (cv::Mat_<double>(3, 3) << 0.866, -0.5, 0, 0.5, 0.866, 0, 0, 0, 1);
    cv::Mat t = (cv::Mat_<double>(3, 1) << 0, 0, 5);
    cv::Point2f P_img(317.08, 263.75);
    double Z = 10.0;

    cv::Mat P_cam = pointImgToCam(K, R, t, P_img, Z);
    std::cout << "转换后的相机坐标为: (" << P_cam.at<double>(0, 0) << ", " << P_cam.at<double>(1, 0) << ", " << P_cam.at<double>(2, 0) << ")" << std::endl;

    return 0;
}

File: python/img_to_cam.py
def point_img_to_cam(K, R, t, P_img, Z):
    K_inv = np.linalg.inv(K)
    Rt_inv = np.linalg.inv(R)
    
    # 计算图像坐标的齐次坐标
    P_img_homo = np.array([P_img[0] * Z, P_img[1] * Z, Z])
    
    # 计算相机坐标系下的坐标
    P_cam = np.dot(Rt_inv, np.dot(K_inv, P_img_homo) - t)
    
    return P_cam

# 测试
K = np.array([[700, 0, 300],
              [0, 700, 200],
              [0, 0, 1]])

R = np.array([[0.866, -0.5, 0],
              [0.5, 0.866, 0],
              [0, 0, 1]])

t = np.array([0, 0, 5])

P_img = np.array([317.08, 263.75])
Z = 10.0

point_img_to_cam(K, R, t, P_img, Z)
